src/Build/Graph/ProjectGraph.cs

```diff
@@ -169,17 +172,27 @@ public ProjectGraph(IEnumerable<ProjectGraphEntryPoint> entryPoints, ProjectColl
         {
             ErrorUtilities.VerifyThrowArgumentNull(projectCollection, nameof(projectCollection));
 
-            var nodeStates = new Dictionary<ConfigurationMetadata, NodeState>();
+            var nodeStates = new Dictionary<ProjectGraphNode, NodeState>();
             var entryPointNodes = new List<ProjectGraphNode>();
+            var tasksInProgress = new ConcurrentDictionary<ConfigurationMetadata, object>();
+            var projectsToEvaluate = new ConcurrentQueue<ConfigurationMetadata>();
+            var entryPointConfigurationMetadata = new List<ConfigurationMetadata>();
             foreach (var entryPoint in entryPoints)
             {
                 PropertyDictionary<ProjectPropertyInstance> globalPropertyDictionary = CreatePropertyDictionary(entryPoint.GlobalProperties);
                 var configurationMetadata = new ConfigurationMetadata(FileUtilities.NormalizePath(entryPoint.ProjectFile), globalPropertyDictionary);
-                if(!nodeStates.TryGetValue(configurationMetadata, out NodeState nodeState))
+                projectsToEvaluate.Enqueue(configurationMetadata);
+                entryPointConfigurationMetadata.Add(configurationMetadata);
+            }
+
+            LoadGraph(projectsToEvaluate, projectCollection, tasksInProgress);
+            foreach(var configurationMetadata in entryPointConfigurationMetadata)
+            {
+                entryPointNodes.Add(_allParsedProjects[configurationMetadata]);
+                if (!nodeStates.TryGetValue(_allParsedProjects[configurationMetadata], out var _))
```

Comment:
Related to Rainer's previous comment, instead of `TryGetValue(key, out var _)` I think `ContainsKey` is even simpler.